crates/bevy_ecs/src/schedule_v3/schedule.rs

```diff
@@ -0,0 +1,1027 @@
+            .unwrap();
+        }
+
+        error!("{}", message);
+    }
+
+    fn contains_cycles(&self, strongly_connected_components: &[Vec<NodeId>]) -> bool {
+        if strongly_connected_components
+            .iter()
+            .all(|scc| scc.len() == 1)
+        {
+            return false;
+        }
+
+        true
+    }
+
+    fn report_cycles(&self, strongly_connected_components: &[Vec<NodeId>]) {
+        let components_with_cycles = strongly_connected_components
+            .iter()
+            .filter(|scc| scc.len() > 1)
+            .cloned()
+            .collect::<Vec<_>>();
+
+        let mut message = format!(
+            "schedule contains at least {} cycle(s)",
+            components_with_cycles.len()
+        );
+
+        writeln!(message, " -- cycle(s) found within:").unwrap();
+        for (i, scc) in components_with_cycles.into_iter().enumerate() {
+            let names = scc
+                .iter()
+                .map(|id| self.get_node_name(id))
+                .collect::<Vec<_>>();
+            writeln!(message, " ---- {i}: {names:?}").unwrap();
+        }
+
+        error!("{}", message);
+    }
+
+    fn contains_conflicts(&self, conflicts: &[(NodeId, NodeId, Vec<ComponentId>)]) -> bool {
+        if conflicts.is_empty() {
+            return false;
+        }
+
+        true
+    }
+
+    fn report_conflicts(&self, ambiguities: &[(NodeId, NodeId, Vec<ComponentId>)]) {
+        let mut string = String::from(
+            "Some systems with conflicting access have indeterminate execution order. \
+            Consider adding `before`, `after`, or `ambiguous_with` relationships between these:\n",
+        );
+
+        for (system_a, system_b, conflicts) in ambiguities {
+            debug_assert!(system_a.is_system());
+            debug_assert!(system_b.is_system());
+            let name_a = self.get_node_name(system_a);
+            let name_b = self.get_node_name(system_b);
+
+            writeln!(string, " -- {name_a} and {name_b}").unwrap();
+            if !conflicts.is_empty() {
+                writeln!(string, "    conflict on: {conflicts:?}").unwrap();
+            } else {
+                // one or both systems must be exclusive
+                let world = std::any::type_name::<World>();
+                writeln!(string, "    conflict on: {world}").unwrap();
+            }
+        }
+
+        warn!("{}", string);
+    }
+}
+
+#[derive(Error, Debug)]
+#[non_exhaustive]
+pub enum BuildError {
+    #[error("`{0:?}` contains itself")]
+    HierarchyLoop(BoxedSystemSet),
+    #[error("system set hierarchy contains cycle(s)")]
+    HierarchyCycle,
+    #[error("system set hierarchy contains conflicting relationships")]
+    HierarchyConflict,
+    #[error("`{0:?}` depends on itself")]
+    DependencyLoop(BoxedSystemSet),
+    #[error("dependencies contain cycle(s)")]
+    DependencyCycle,
+    #[error("`{0:?}` and `{1:?}` can have a hierarchical OR dependent relationship, not both")]
```

Comment:
I think using the term "hierarchical" on its own will be confusing for users. It would be helpful to contextualize this a bit with terminology users actually use in their apps ("hierarchy" relates to "set", "dependent" relates to  "before" or "after")

Reply:
So should the error message be this then?

> `A` and `B` can have an `in_set` OR `before`/`after` relationship, not both.

Reply:
Ideally we lay this out a bit more, ala the rustc errors.

> `A` is in the same set as `B`, but it is also [before/after] all systems in `B`
> 
> This results in a broken schedule, as systems cannot run before themselves.

Maybe we customize this based on ordering type, maybe we reword.

Reply:
`in_set` -> `contains`?